File: maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.java
class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int len = s.length(), range[][] = new int[26][2];
        for (int i = 0; i < 26; i++) range[i] = new int[]{len, 0};
        Stack<int[]> st = new Stack<>();
        List<String> res = new ArrayList<>();
        for (int i = 0; i < len; i++) {
            int idx = s.charAt(i) - 'a';
            range[idx][0] = Math.min(i, range[idx][0]);
            range[idx][1] = Math.max(i, range[idx][1]);
        }
        for (int i = 0; i < len; i++) {
            int idx = s.charAt(i) - 'a';
            if (range[idx][0] != i) continue;
            int l = range[idx][0], tail = range[idx][1], r = getRightMost(l, tail, range, s);
            if (r < 0) continue;
            while (!st.isEmpty() && l >= st.peek()[0] && r <= st.peek()[1]) st.pop();
            st.push(new int[]{l, r});
        }
        while (!st.isEmpty()) res.add(s.substring(st.peek()[0], st.pop()[1] + 1));
        return res;
    }
    
    private int getRightMost(int l, int r, int[][] range, String s){
        for (int i = l + 1; i < r; i++) {
            int idx = s.charAt(i) - 'a';
            if (range[idx][0] < l) return -1;
            r = Math.max(r, range[idx][1]);
        }
        return r;
    }
}

File: maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.cs

public class Solution {
    public IList<string> MaxNumOfSubstrings(string s) {
        int[] low = new int[26];
        int[] high = new int[26];
        Array.Fill(low, s.Length);
        for(int i = 0; i < s.Length; i++) {
            int index = s[i] - 'a';
            low[index] = Math.Min(low[index], i);
            high[index] = Math.Max(high[index], i);
        }
        var stack = new Stack<int[]>();
        for(int i = 0; i < s.Length; i++) {
            int index = s[i] - 'a';
            int l = low[index], r = high[index];
             Console.WriteLine("i " + i  + " l " + l);
            if(i != l) {
                continue;
            }
           
            int end = DoubleCheckHigh(s, l, r, low, high);
            Console.WriteLine("i " + i  + " end " + end);
            if(end < 0) {
                continue;
            }
            while(stack.Count > 0 && l >= stack.Peek()[0] 
                  && end <= stack.Peek()[1]) {
                Console.WriteLine("pop " + stack.Peek()[0]  + " " + stack.Peek()[1]);
                stack.Pop();
            }
            Console.WriteLine("push " + l  + " " + end);
            stack.Push(new int[]{l, end});
        }
        IList<string> lst = new List<string>();
        while(stack.Count > 0) {
            int[] range = stack.Pop();
            lst.Add(s.Substring(range[0], range[1] - range[0] + 1));
        }
        return lst;
    }
    private int DoubleCheckHigh(string s, int l, int r, int[] low, int[] high) {
        for(int i = l + 1; i < r; i++) {
            if(low[s[i] - 'a'] < l) {
                return -1;
            }
            r = Math.Max(high[s[i] - 'a'], r);
        }
        return r;
    }
}